lib/charms/istio_k8s/v0/istio_info.py

```diff
@@ -0,0 +1,239 @@
+
+        raw_data = relations[0].data.get(relations[0].app)
+        if raw_data == {} and skip_empty:
+            return None
+
+        return IstioInfoAppData.load(relations[0].data[relations[0].app])
+
+    def get_data_from_all_relations(self, skip_empty: bool = True) -> List[IstioInfoAppData]:
+        """Return a list of data objects from all relations.
+
+        Args:
+            skip_empty: If True, return None if there is a relation that has not yet provided any data.
+                        If False, raise a DataValidationError
+                        TODO: is it practical for this to ever be used with skip_empty=False?  Should it be removed?
+        """
+        info_list = []
+        for relation in self._get_relations():
+            data = relation.data.get(relation.app)
+            if data == {} and skip_empty:
+                continue
+            info_list.append(IstioInfoAppData.load(relation.data[relation.app]))
+        return info_list
+
+
+# Provider library
+
+class SendInfoFailedEvent(EventBase):
+    """Charm Event triggered when the istio-info provider fails to send data successfully."""
+
+
+class IstioInfoProviderCharmEvents(CharmEvents):
+    """Events raised by the IstioInfoProvider class."""
+    # TODO: If this becomes frequently used, consider adding more data to the event
+    send_info_failed = EventSource(SendInfoFailedEvent)
+
+
+class IstioInfoProvider(Object):
+    """Class for handling the provider side of the istio-info relation."""
+
+    on = IstioInfoProviderCharmEvents()
+
+    def __init__(
+            self,
+            charm: CharmBase,
+            root_namespace: str,
+            relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        """Initialize the IstioInfoProvider object.
+
+        Args:
+            charm: The charm instance.
+            root_namespace: The root namespace for the Istio installation.
+            relation_name: The name of the relation.
+        """
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._root_namespace = root_namespace
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_joined,
+            self.handle_send_data_event
+        )
+        # Observe leader elected events because only the leader should send data, and we don't want a case where the
+        # relation_joined event happens during a leadership change.
+        self.framework.observe(
+            self._charm.on.leader_elected,
+            self.handle_send_data_event
+        )
+
+    def handle_send_data_event(self, _: EventBase) -> None:
+        """Handle events that should send data to the relation."""
+        try:
+            if self._charm.unit.is_leader():
+                self.send_data()
+        except Exception:
+            # TODO: This doesn't work!  Syntax is wrong
+            self.on.send_info_failed.emit()
```

Comment:
Do we even need this?  The only times this happens are if we are not a leader (which we guard against) and if something really bad happens with juju.  In that case, having unhandled exception feels correct. 

My inclination is to get rid of this.

Reply:
I agree, if an exception happens here it is a mysterious error and should just be bubbled up as usual. Emitting a useless event is not really "handling the exception".

Reply:
Removed